File: frontend/src/api/adherence.ts
import api from './client';

export interface MealTypeStats {
  meal_type: string;
  completed: number;
  skipped: number;
  deviated: number;
}

export interface DailyAdherence {
  date: string;
  completed: number;
  skipped: number;
  deviated: number;
  adherence_pct: number;
}

export interface RecentMealLog {
  log_date: string;
  meal_type: string;
  status: string;
  deviation_note?: string | null;
  logged_at?: string | null;
}

export interface ClientAdherence {
  client_id: string;
  period_days: number;
  total_completed: number;
  total_skipped: number;
  total_deviated: number;
  adherence_pct: number;
  daily: DailyAdherence[];
  by_meal_type: MealTypeStats[];
  recent_logs: RecentMealLog[];
}

export const adherenceApi = {
  getClientAdherence: async (clientId: string, days = 7) => {
    const response = await api.get<ClientAdherence>(`/clients/${clientId}/adherence`, {
      params: { days },
    });
    return response.data;
  },
};

File: backend/app/schemas/adherence.py
"""Pydantic schemas for meal adherence stats."""

from datetime import date, datetime

from pydantic import BaseModel


class MealTypeStats(BaseModel):
    meal_type: str
    completed: int
    skipped: int
    deviated: int


class DailyAdherence(BaseModel):
    date: date
    completed: int
    skipped: int
    deviated: int
    adherence_pct: float


class RecentMealLog(BaseModel):
    log_date: date
    meal_type: str
    status: str
    deviation_note: str | None = None
    logged_at: datetime | None = None


class ClientAdherenceResponse(BaseModel):
    client_id: str
    period_days: int
    total_completed: int
    total_skipped: int
    total_deviated: int
    adherence_pct: float
    daily: list[DailyAdherence]
    by_meal_type: list[MealTypeStats]
    recent_logs: list[RecentMealLog]


class AttentionClient(BaseModel):
    id: str
    name: str
    adherence_pct: float
    last_interaction: date | None = None


class RecentActivity(BaseModel):
    type: str
    client: str
    timestamp: datetime
    detail: str | None = None


class DashboardOverview(BaseModel):
    total_clients: int
    active_clients: int
    plans_this_month: int
    pending_approvals: int
    avg_adherence_pct: float
    clients_needing_attention: list[AttentionClient]
    recent_activity: list[RecentActivity]
